File: AlerteTCL/Services/BusTrackerService.swift
import Foundation

/// Récupère les informations de véhicule depuis l'API publique Bus Tracker.
/// Source externe non officielle — données manquantes ou erreurs silencieusement ignorées.
/// Référence : https://bus-tracker.fr
enum BusTrackerService {
    private static let baseURL = "https://bus-tracker.fr/api/vehicles"
    private static let networkId = 91 // TCL Lyon (id stable dans la DB Bus Tracker)

    private static var cache: [String: String] = [:]

    static func fetchVehicleModel(fleetNumber: String) async -> String? {
        if let cached = cache[fleetNumber] { return cached }
        guard let url = URL(string: "\(baseURL)?networkId=\(networkId)&number=\(fleetNumber)") else { return nil }
        var request = URLRequest(url: url, timeoutInterval: 8)
        request.setValue("AlerteTCL/1.0", forHTTPHeaderField: "User-Agent")
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return nil }
            let vehicles = try JSONDecoder().decode([BusTrackerVehicleDTO].self, from: data)
            guard let designation = vehicles.first(where: { $0.number == fleetNumber })?.designation else { return nil }
            cache[fleetNumber] = designation
            return designation
        } catch {
            return nil
        }
    }

    private struct BusTrackerVehicleDTO: Decodable {
        let number: String?
        let designation: String?
    }
}

File: shared/src/commonMain/kotlin/com/alertetcl/shared/services/BusTrackerService.kt
package com.alertetcl.shared.services

import com.alertetcl.shared.network.HttpClientProvider
import io.ktor.client.call.body
import io.ktor.client.plugins.timeout
import io.ktor.client.request.get
import io.ktor.http.HttpStatusCode
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable

/**
 * Récupère les informations de véhicule depuis l'API publique Bus Tracker.
 * Source externe non officielle — données manquantes ou erreurs silencieusement ignorées.
 * Référence : https://bus-tracker.fr
 */
class BusTrackerService {
    private val client = HttpClientProvider.client
    private val baseURL = "https://bus-tracker.fr/api/vehicles"
    private val networkId = 91 // TCL Lyon (id stable dans la DB Bus Tracker)

    private val mutex = Mutex()
    private val cache = mutableMapOf<String, String>()

    /** Retourne la désignation (modèle) du véhicule, ou null si non trouvé / erreur. */
    suspend fun fetchVehicleModel(fleetNumber: String): String? {
        mutex.withLock { cache[fleetNumber] }?.let { return it }
        return try {
            val resp = client.get("$baseURL?networkId=$networkId&number=$fleetNumber") {
                timeout { requestTimeoutMillis = 8_000L }
            }
            if (resp.status != HttpStatusCode.OK) return null
            val vehicles: List<BusTrackerVehicleDto> = resp.body()
            val designation = vehicles.firstOrNull { it.number == fleetNumber }?.designation ?: return null
            mutex.withLock { cache[fleetNumber] = designation }
            designation
        } catch (_: Throwable) {
            null
        }
    }

    @Serializable
    private data class BusTrackerVehicleDto(val number: String? = null, val designation: String? = null)

    companion object { val shared = BusTrackerService() }
}
